ext/drreg/drreg.c

```diff
@@ -500,16 +500,16 @@ drreg_insert_restore_all(void *drcontext, instrlist_t *bb, instr_t *inst,
     /* Before each app read, or at end of bb, restore spilled registers to app values.
      * XXX i#5121: For regs that are in spilled state at this instr also during the
      * app2app phase, the following does not restore the app value, it restores the
-     * app2app meta value instead. This is expected and correct, because the drreg use
-     * in insertion phase adds spills/restores around the drreg spills/restores from
-     * previous phases like they were app instrs. So, the app2app meta value is no
-     * different than the app value for the insertion phase.
+     * app2app meta value instead. This is expected and correct, because drreg in
+     * insertion phase adds spills/restores around the drreg spills/restores from the
+     * previous phases like they were app instrs. So, the app2app meta value is treated
+     * like an app value by the insertion phase.
```

Comment:
I still find this confusing, in the order it talks about things and the "expected and correct" for the non-ideal case of not having app2app restore.  This is the place in the code where drreg is adding spills/restores, but the comment implies it's somewhere else.

I would suggest:

XXX i#5121: Any app2app cross-app-instr use should manually restore app values prior to app reads.  That interacts well with the insertion phase since we treat those app2app restores as app writes and end up with the app value here.  If the app2app phase has a regsiter in a spilled state at this instr and did not restore the app value, the following insertion phase code will not restore the app value but instead will restore the app2app meta value.

Reply:
Done. Thanks for the detailed suggestion! Couldn't figure out the best words here on my own!